### DISCUSSION TITLE
[New] Tree sitter for zig

### DISCUSSION THREAD

**Discussion by maxxnino** [tree-sitter/tree-sitter#1313] [Grammars]
## Info
> I'm working on new tree sitter parse for zig, using official [zig-spec](https://github.com/ziglang/zig-spec) repo. Everything seem to working really good.
The generated c files only 3,6m which is really small,.
I need help for contribution(more test, highlight, local...)
My repo [tree-sitter-zig](https://github.com/maxxnino/tree-sitter-zig)

## Progress

- [X] Highlight
- [x] Indent
- [x] Fold
- [ ] Local

## Image

![demo](https://user-images.githubusercontent.com/34153891/128273994-e3a3cb7b-12e8-437b-8bb4-7ffc7f81a894.png)

**Comment by ahlinc** [comment]
I converted this to a discussion.
_By the way, good job :+1: It would be awesome to see does it possible to define the same Zig grammar that is complete and has a small generated parser._

**Comment by maxxnino** [comment]


**Reply by ahlinc** [reply]
> When writing a query I have two scenario:
> 
> Scenario 1: if a (SuffixOp) node don't have (FnCallArguments) node below it. The (IDENTIFIER) node become a field
> Scenario 2: if a (SuffixOp) node have, The (IDENTIFIER) node become a function
> How can i achive that?
> ```
> (ErrorUnionExpr
>   ; Scenario 1
>   (SuffixOp (IDENTIFIER) @field)
> 
>   ; Scenario 2
>   (SuffixOp (IDENTIFIER) @function)
>   (FnCallArguments)
> 
>   ; Scenario 1
>   (SuffixOp (IDENTIFIER) @field)
> 
>   ; Scenario 2
>   (SuffixOp (IDENTIFIER) @function)
>   (FnCallArguments)
> )
> ```

Try something like this with [a dot anchor](https://tree-sitter.github.io/tree-sitter/using-parsers#query-syntax) and split your query to two queries:
```
; Scenario 1
(ErrorUnionExpr
  (SuffixOp (IDENTIFIER) @field)
  .
)

; Scenario 2
(ErrorUnionExpr
  (SuffixOp (IDENTIFIER) @function)
  (FnCallArguments)
)
```

**Reply by maxxnino** [reply]
Sorry, I accidently delete the comment above. Btw, I re written this section, make it easier  for query. Don't have the opportunity to test your suggestion yet. I'll try in the future. Thank you for the infomation

**Reply by ahlinc** [reply]
No worries, it was easy to find in a mailbox what was there, I've put it as a quote in my previous answer.

**Reply by ahlinc** [reply]
@maxxnino Could you tell what is an editor on your screenshot in the description that allows such icons on the left?

**Reply by maxxnino** [reply]
@ahlinc I'm using neovim with [LunarVim](https://github.com/LunarVim/LunarVim), and [todo comment](https://github.com/folke/todo-comments.nvim)

**Reply by maxxnino** [reply]
I tried your query suggestion, and it's working very good

![Untitled](https://user-images.githubusercontent.com/34153891/128623332-9b030ca7-a1b2-4c18-ab47-a3d282621aa4.jpg)
```scm
; var.field
(SuffixOp (IDENTIFIER) @field)

; var.func().filed.func().field.filed.func()
( 
  (SuffixOp (IDENTIFIER) @function)
  .
  (FnCallArguments)
)
; func()
( 
  ((IDENTIFIER) @function)
  .
  (FnCallArguments)
)

```

**Comment by maxxnino** [comment]
@ahlinc could you help me with this [issue](https://github.com/maxxnino/tree-sitter-zig/issues/3). How to make it prefer `IfStatement` instead if `AssignExpr`

**Reply by ahlinc** [reply]
I'm sorry, I'm not familiar with the Zig spec and currently don't have enough of the free time for this.
But I've briefly looked on your grammar and I think the issue is in a `SuffixExpr` rule, I dont see a construct in it that would be able to consume the `aa.bb` sequence. Also the `ERROR` happens later than expected cause you have defined `conflicts` that allows tree-sitter to use a GLR algorithm and try another sequence of tokens by applying an another GLR branch. I'd suggest to minimize amount of defined `conflicts` as much as possible, because having a lot of defined conflicts may have a win in a generated grammar size but loose in it's overall performance, but that's just is my assumption. Also using the `conflicts` leads to situation that tree-sitter looks over GLR branches of alternative terminal sequences and it seems leaves the `ERROR` from the last checked branch that also were failed and this leads to the ERROR nodes in less expected places.

**Reply by maxxnino** [reply]
`SuffixExpr` indeed can comsume the aa.bb sequence with ` seq($._PrimaryTypeExpr, repeat(choice($.SuffixOp, $.FnCallArguments)))`.
The `_PrimaryTypeExpr` can expand to `IDENTIFIER` and `SuffixOp` can expand to `seq(DOT, $.IDENTIFIER)`.

I tried minimize amount of defined `conflicts` to zero. And tried prec.left/right combination. But nothing seem to beat the current grammar. Maybe. i don't understand how `prec.left` and `prec.right` work. In the [using-associativity](https://tree-sitter.github.io/tree-sitter/creating-parsers#using-associativity) example. Does that mean `prec.right` always the number 1, and `prec.left` always the number 2

**Reply by maxxnino** [reply]
Finally, I fixed all parsing error, and solved all conflicts https://github.com/maxxnino/tree-sitter-zig/commit/4f4f96b9dc53fee7cc6287712714a2a5c6f93c2a

**Reply by ahlinc** [reply]
Awesome work! I see you've avoided to use the `conglicts` at all :tada:

**Comment by maxxnino** [comment]
@ahlinc how can i test parser performance? I only known two way, one is c generated file which is 3.7mb. The other is using command `tree-sitter generate  --report-states-for-rule  -`
